src/v/cluster/members_manager.cc

```diff
@@ -328,20 +328,17 @@ wait_for_next_join_retry(std::chrono::milliseconds tout, ss::abort_source& as) {
       });
 }
 
-ss::future<result<join_reply>> members_manager::dispatch_join_to_remote(
-  const config::seed_server& target, model::broker joining_node) {
+ss::future<result<join_node_reply>> members_manager::dispatch_join_to_remote(
+  const config::seed_server& target, join_node_request&& req) {
     vlog(clusterlog.info, "Sending join request to {}", target.addr);
     return do_with_client_one_shot<controller_client_protocol>(
       target.addr,
       _rpc_tls_config,
       _join_timeout,
-      [joining_node = std::move(joining_node),
-       timeout = rpc::clock_type::now()
-                 + _join_timeout](controller_client_protocol c) mutable {
-          return c
-            .join(
-              join_request(std::move(joining_node)), rpc::client_opts(timeout))
-            .then(&rpc::get_ctx_data<join_reply>);
+      [req = req, timeout = rpc::clock_type::now() + _join_timeout](
```

Comment:
do we need `req= std::move(req)` ?

Reply:
I think you're right.  We were getting away with this if do_with_client_on_shot was not hitting a scheduling point before executing this lambda (the inner lambda moves), but there should be a move here to make it correct in the general case.  Added.